core/vm/actor/storage_power_actor.cpp

```diff
@@ -0,0 +1,228 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "vm/actor/storage_power_actor.hpp"
+#include <power/impl/power_table_impl.hpp>
+#include "vm/indices/indices.hpp"
+
+namespace fc::vm::actor {
+
+  const power::Power StoragePowerActor::kMinMinerSizeStor =
+      100 * (primitives::BigInt(1) << 40);
+
+  const size_t StoragePowerActor::kMinMinerSizeTarg = 3;
+
+  outcome::result<std::vector<primitives::address::Address>>
+  StoragePowerActor::selectMinersToSurprise(
+      size_t challenge_count,
+      const crypto::randomness::Randomness &randomness) {
+    std::vector<primitives::address::Address> selected_miners;
+
+    if (power_table_->getSize() < challenge_count) return OUT_OF_BOUND;
+
+    OUTCOME_TRY(all_miners, power_table_->getMiners());
+
+    if (challenge_count == all_miners.size()) return std::move(all_miners);
+
+    for (size_t chall = 0; chall < challenge_count; chall++) {
+      int miner_index =
+          randomness_provider_->randomInt(randomness, chall, all_miners.size());
+      auto potential_challengee = all_miners[miner_index];
+
+      all_miners.erase(all_miners.begin() + miner_index);
+
+      selected_miners.push_back(potential_challengee);
+    }
+
+    return selected_miners;
+  }
+
+  outcome::result<void> StoragePowerActor::addClaimedPowerForSector(
+      const primitives::address::Address &miner_addr,
+      const SectorStorageWeightDesc &storage_weight_desc) {
+    // TODO(artyom-yurin): FROM SPEC: The function is located in the indices
```

Comment:
please refer to a task with the corresponding description

Reply:
Fixed